server/config/component_config.go

```diff
@@ -0,0 +1,494 @@
+		} else {
+			c.LocalCfgs[component][componentID] = lc
+			status = &configpb.Status{Code: configpb.Status_OK}
+		}
+	}
+
+	config, err := c.getComponentCfg(component, componentID)
+	if err != nil {
+		status = &configpb.Status{Code: configpb.Status_UNKNOWN, Message: "encode error"}
+		return latestVersion, "", status
+	}
+
+	return latestVersion, config, status
+}
+
+func (c *ComponentsConfig) getLatestVersion(component, componentID string) *configpb.Version {
+	v := &configpb.Version{
+		Global: c.GlobalCfgs[component].GetVersion(),
+		Local:  c.LocalCfgs[component][componentID].GetVersion().GetLocal(),
+	}
+	return v
+}
+
+func (c *ComponentsConfig) getComponentCfg(component, componentID string) (string, error) {
+	config := c.LocalCfgs[component][componentID].GetConfigs()
+	updateEntries := make(map[string]*EntryValue)
+	// apply the global change to updateEntries
+	if _, ok := c.GlobalCfgs[component]; ok {
+		globalUpdateEntries := c.GlobalCfgs[component].GetUpdateEntries()
+		for k, v := range globalUpdateEntries {
+			updateEntries[k] = v
+		}
+	}
+	// apply the local change to updateEntries
+	for k1, v1 := range c.LocalCfgs[component][componentID].GetUpdateEntries() {
+		if v, ok := updateEntries[k1]; ok {
+			// apply conflict
+			if v1.Version.GetGlobal() == v.Version.GetGlobal() {
+				updateEntries[k1] = v1
+			}
+		} else {
+			updateEntries[k1] = v1
+		}
+	}
+
+	for k, v := range updateEntries {
+		configName := strings.Split(k, ".")
+		update(config, configName, v.Value)
+	}
+
+	return encodeConfigs(config)
+}
+
+// Update ...
+func (c *ComponentsConfig) Update(kind *configpb.ConfigKind, version *configpb.Version, entries []*configpb.ConfigEntry) (*configpb.Version, *configpb.Status) {
+	c.Lock()
+	defer c.Unlock()
+
+	global := kind.GetGlobal()
+	if global != nil {
+		component := global.GetComponent()
+		// if the global config of the component is existed.
+		if globalCfg, ok := c.GlobalCfgs[component]; ok {
+			globalLatestVersion := globalCfg.GetVersion()
+			if globalLatestVersion > version.GetGlobal() {
+				return &configpb.Version{Global: globalLatestVersion, Local: version.GetLocal()},
+					&configpb.Status{
+						Code:    configpb.Status_STALE_VERSION,
+						Message: "version is wrong",
+					}
+			} else if globalLatestVersion < version.GetGlobal() {
+				return &configpb.Version{Global: globalLatestVersion, Local: version.GetLocal()},
+					&configpb.Status{
+						Code:    configpb.Status_UNKNOWN,
+						Message: "version is wrong",
+					}
+			}
+
+			newGlobalVersion := version.GetGlobal() + 1
+			for _, entry := range entries {
+				globalCfg.updateEntry(entry, &configpb.Version{Global: newGlobalVersion, Local: 0})
+			}
+			globalCfg.Version = newGlobalVersion
+			// update all local config version
+			for _, LocalCfg := range c.LocalCfgs[component] {
+				LocalCfg.Version = &configpb.Version{Global: newGlobalVersion, Local: 0}
+			}
```

Comment:
Maybe we don't need to change the local version here, if there is no priority relation between global and local version.

Reply:
The purpose of it is that I want to let the local config know the global config is valid by changing the local version to `0`

Reply:
The version field on response contains both global and local version, which can also  that work. Moreover, by separating the update of global and local version, the client can known exactly which kind of config had changed recently.